underdocs-renderer/src/main/kotlin/underdocs/renderer/output/html/render/section/SectionRenderer.kt

```diff
@@ -310,31 +311,49 @@ class SectionRenderer(private val linker: Linker) {
             h2("See Also")
     ).withClass("see-also")
 
-    fun renderReturnValue(visitable: Visitable, returnValue: String): Tag<*> {
-        val elements = ArrayList<DomContent>()
-
-        elements.add(h2("Return Value"))
-
-        if (visitable is Function) {
-            elements.add(p(
-                    code(referredTypeToString(visitable.returnType))
-            ).withClass("return-type"))
-        }
-
-        elements.add(renderMarkdown(returnValue))
-
-        return section().with(elements).withClass("return-value")
-    }
-
-    private fun referredTypeToString(type: ReferredType): String {
-        var representation = type.specifiers.joinToString(" ")
-
-        if (!representation.endsWith("*")) {
-            representation += " "
-        }
-
-        return representation + type.name
-    }
+    fun renderReturnValue(returnValue: ReturnValue) = section(
+        h2("Return Value"), div(
+            h3("Success"), table(
+                tbody(
+                    each(returnValue.success) { child ->
+                        tr(
+                            td(
+                                h4(child.value)
+                            ).withClass("element-name-cell"),
+                            td(
+                                div(
+                                    renderMarkdown(child.description)
+                                ).withClass("element-excerpt-text"),
+                                div(
+                                    "type info comes here"
+                                ).withClass("element-excerpt-type")
+                            ).withClass("element-excerpt-cell")
+                        )
+                    }
+                )
+            ).withClass("elements-table")
+        ), div(
```

Comment:
Note, that this `div` is rendered even if there are no errors in the `Return Value` section.

To resolve this situation, I see two possible solutions.

**Incremental DOM Building**

Turn the expression-bodied function into a normal one. Then, create the `section` node with the `h2` and continue by add the two `div`s incrementally, and checking whether there are error value.

**Expression-body with if Function**

Keep the expression-bodied function. Also, create a new function with a signature like 
~~~~Kotlin
fun <T> iffPrimitive(condition: Boolean, ifFunction: () -> T?): T?
~~~~

(see [j2html TagCreator.iff](https://github.com/tipsy/j2html/blob/master/src/main/java/j2html/TagCreator.java#L37) regarding the implementation).

Then, you can do:

~~~~Kotlin
iffPrimitive(returnValue.error.isNotEmpty()) {
  div(...)
}
~~~~

(see [Trailing Lambda syntax](https://kotlinlang.org/docs/reference/lambdas.html#passing-a-lambda-to-the-last-parameter))

Although `j2html` already provides `iff` functions for scenarios like this, they are far from perfect:

  * The one with signature `Boolean, T` is not lazy: `T` is always produced. By using `() -> T`, we only produce `T` if the condition evaluates to `true`.
  * The other one is `Optional<U>, U -> T`. Sure, we could create an optional and then used the identity function as a converter, but that's plain ugly.

Another option would've been the use of an [extension function](https://kotlinlang.org/docs/reference/extensions.html#extension-functions) on `TagCreator`. However, because `TagCreator` is written in Java, we cannot write

~~~~Kotlin
fun <T> TagCreator.Companion.iffPrimitive
~~~~

but only

~~~~Kotlin
fun <T> TagCreator.iffPrimitive
~~~~

which makes no sense,

Reply:
Yup, good stuff going on in the `renderReturnValue` function, simply great!

One thing, I'd add though: can you please move the `iffPrimitive` function elsewhere? The `SectionParserUtils` file is mostly concerned with Markdown parse tree navigation and manipulation, while `iffPrimitive` is more of an HTML-rendering stuff.

Reply:
I've created a new file `RendererHelper` in the underdocs-renderer module.